Sources/Memcache/MemcacheConnection.swift

```diff
@@ -56,23 +67,63 @@ public actor MemcacheConnection: Service {
 
     private var state: State
 
-    /// Initialize a new MemcacheConnection.
+    /// Initialize a new MemcacheConnection, with an option to specify an ID.
+    /// If no ID is provided, a default value is used.
     ///
     /// - Parameters:
     ///   - host: The host address of the Memcache server.
     ///   - port: The port number of the Memcache server.
     ///   - eventLoopGroup: The event loop group to use for this connection.
-    public init(host: String, port: Int, eventLoopGroup: EventLoopGroup) {
+    ///   - id: The unique identifier for the connection (optional).
+    public init(host: String, port: Int, id: ID? = nil, eventLoopGroup: EventLoopGroup) {
         self.host = host
         self.port = port
+        self.id = id ?? MemcacheConnection.nextID.wrappingIncrementThenLoad(ordering: .sequentiallyConsistent)
         let (stream, continuation) = AsyncStream<StreamElement>.makeStream()
         let bufferAllocator = ByteBufferAllocator()
-        self.state = .initial(
-            eventLoopGroup: eventLoopGroup,
-            bufferAllocator: bufferAllocator,
-            requestStream: stream,
-            requestContinuation: continuation
-        )
+        self.closePromise = eventLoopGroup.next().makePromise(of: Void.self)
+        self.state = .initial(eventLoopGroup: eventLoopGroup, bufferAllocator: bufferAllocator, requestStream: stream, requestContinuation: continuation)
+    }
+
+    deinit {
+        // Fulfill the promise if it has not been fulfilled yet
+        closePromise.fail(MemcacheError(code: .connectionShutdown,
+                                        message: "MemcacheConnection deinitialized without closing",
+                                        cause: nil,
+                                        location: .here()))
+    }
+
+    /// Closes the connection. This method is responsible for properly shutting down
+    /// and cleaning up resources associated with the connection.
+    public nonisolated func close() {
+        Task {
+            await self.closeConnection()
+        }
+    }
+
+    private func closeConnection() async {
+        switch self.state {
+        case .running(_, let channel, _, _):
+            channel.channel.close().cascade(to: self.closePromise)
+        default:
+            self.closePromise.succeed(())
+        }
+        self.state = .finished
+    }
+
+    /// Registers a closure to be called when the connection is closed.
+    /// This is useful for performing cleanup or notification tasks.
+    public nonisolated func onClose(_ closure: @escaping ((any Error)?) -> Void) {
+        Task {
+            await self.closeFuture.whenComplete { result in
```

Comment:
if the closeFuture is marked as `nonisolated` we don't need a Task and an await here.